<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <link href="//fonts.googleapis.com/css?family=Raleway:400,300,600" rel="stylesheet" type="text/css">
  <link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/skeleton/2.0.4/skeleton.css"/>

  <!-- Necessary includes for LocusZoom.js -->
 <script src="https://cdn.jsdelivr.net/npm/d3@^5.16.0" type="text/javascript"></script>
  <script src="../dist/locuszoom.app.min.js" type="text/javascript"></script>
  <script type="application/javascript" src="../dist/ext/lz-dynamic-urls.min.js"></script>

  <link rel="stylesheet" href="../dist/locuszoom.css" type="text/css"/>

  <script type="application/javascript" src="../dist/ext/lz-tabix-source.min.js"></script>
  <script src="js/coaccessibility.js" type="application/javascript"></script>

  <title>LocusZoom.js ~ Coaccessibility</title>

  <style>
    body {
      background-color: #FAFAFA;
      margin: 0px 20px;
    }

    img {
      max-width: 100%;
      box-sizing: border-box;
    }
  </style>

</head>

<body>
<div class="container">

  <h1 style="margin-top: 1em;"><strong>LocusZoom.js</strong></h1>

  <h3 style="float: left; color: #777">Chromatin Co-accessibility</h3>
  <h6 style="float: right;"><a href="../index.html">&lt; return home</a></h6>

  <hr style="clear: both;">
  <p>
    A LocusZoom extension track for interactively visualizing single-cell chromatin co-accessibility links
    & scores of sites in genome to predict cis-regulatory interactions.
  </p>

  <p>
    Data from: Rai, V. et al. Single-cell ATAC-Seq in human pancreatic islets and deep learning upscaling of rare cells reveals cell-specific type 2 diabetes regulatory signatures. Mol Metab 32, 109-121 (2020). (PMID <a href="https://pubmed.ncbi.nlm.nih.gov/32029221/">32029221</a>)
  </p>

  <p>
    This demonstration shows how to read annotations directly from an external BED track (tabix indexed), without
    needing to convert the data into JSON or load via a remote API. The remote file server must enable some basic
    headers. (see <a
      href="See: https://docs.cancergenomicscloud.org/docs/enabling-cross-origin-resource-sharing-cors#CORS">instructions
    for google cloud and s3 buckets</a>)
  </p>

  <p>
    <a href="https://github.com/statgen/locuszoom/blob/master/examples/coaccessibility.html">View source code</a>
  </p>

  <div id="lz-plot" class="lz-container-responsive"></div>
  <hr>

  <div class="row">
    <footer style="text-align: center;">
      &copy; Copyright
      <script>document.write(new Date().getFullYear())</script>
      <a href="https://github.com/statgen">The University of Michigan Center for Statistical Genetics</a><br>
    </footer>
  </div>
</div>

<script type="text/javascript">
    var apiBase = "https://portaldev.sph.umich.edu/api/v1/";
    data_sources = new LocusZoom.DataSources()
        .add("assoc", ["AssociationLZ", {url: apiBase + "statistic/single/", source: 45 }])
        .add("ld", ["LDServer", { url: "https://portaldev.sph.umich.edu/ld/", source: '1000G', build: 'GRCh37', population: 'ALL' }])
        .add("recomb", ["RecombLZ", { url: apiBase + "annotation/recomb/results/", build: 'GRCh37' }])
        .add("access", ["TabixUrlSource", {
            // Corresponds to: https://www.diabetesepigenome.org/files/DFF044MQE/
            //   as processed from https://www.diabetesepigenome.org/annotations/DSR299XDW/
            // Islet of Langerhans / Chicago analysis, from CROCPOT
            url_data: 'https://locuszoom.s3-us-west-2.amazonaws.com/DFF044MQE_sorted.bed.gz',
            parser_func: dega_bed_parser,
            // We are fetching two elements, but only one of them is captured in the tabix index.
            // Fetching 25% extra data on each side can ensure that things near the edge are returned when we query.
            overfetch: 0.25,
        }])
        .add("gene", ["GeneLZ", { url: apiBase + "annotation/genes/", build: 'GRCh37' }])
        .add("constraint", ["GeneConstraintLZ", { url: "https://gnomad.broadinstitute.org/api/", build: 'GRCh37' }]);

    // Get the standard association plot layout from LocusZoom's built-in layouts
    var stateUrlMapping = { chr: "chrom", start: "start", end: "end", ldrefvar: 'ld_variant' };
    // Fetch initial position from the URL, or use some defaults
    var initialState = LzDynamicUrls.paramsFromUrl(stateUrlMapping);
    if (!Object.keys(initialState).length) {
        initialState = { chr: '10', start: 114176406, end: 115176406 };
    }
    layout = LocusZoom.Layouts.get("plot", "coaccessibility", { state: initialState });
    LocusZoom.Layouts.mutate_attrs(layout, '$.panels[?(@.tag === "coaccessibility")]', (config) => {
        config.title = { text: 'Pancreatic Islet alpha cells from snATAC-seq' };
        config.toolbar.widgets.push(
            {
                type: 'filter_field',
                position: 'right',
                layer_name: 'coaccessibility',
                field: 'access:score',
                field_display_html: 'Score',
                operator: '>=',
                data_type: 'number',
            }
        );
        return config;
    });


    // Generate the LocusZoom plot, and reflect the initial plot state in url
    window.plot = LocusZoom.populate("#lz-plot", data_sources, layout);

    // Changes in the plot can be reflected in the URL, and vice versa (eg browser back button can go back to
    //   a previously viewed region)
    LzDynamicUrls.plotUpdatesUrl(plot, stateUrlMapping);
    LzDynamicUrls.plotWatchesUrl(plot, stateUrlMapping);
</script>
</body>
</html>
